SYNC.md

```diff
@@ -0,0 +1,134 @@
+
+Here is a simple example:
+```
+function updatePeers(docId: string) {
+  Object.entries(syncStates).forEach(([peer, syncState]) => {
+    const [nextSyncState, syncMessage] = Backend.generateSyncMessage(
+      backends[docId],
+      syncState[docId] || Backend.initSyncState(),
+    )
+    syncStates[peer] = { ...syncStates[peer], [docId]: nextSyncState }
+    if (syncMessage) {
+      sendMessage({
+        docId, source: workerId, target: peer, syncMessage,
+      })
+    }
+  })
+}
+```
+### Receiving sync messages
+Receiving sync messages is also simple. Just pass the document, syncState, and incoming message to `receiveSyncMessage()`, and keep the updated results. Calling `receiveSyncMessage` may also produce a `patch`, which you must forward to the frontend.
+
+After receiving a sync message, you should check if you need to send new sync messages to any connected peers using the code above. In our example code below this is represented by a call to `updatePeers()`
+
+```
+  const [nextBackend, nextSyncState, patch] = Backend.receiveSyncMessage(
+    backends[docId],
+    syncStates[source][docId] || Backend.initSyncState(),
+    syncMessage,
+  )
+  backends[docId] = nextBackend
+  syncStates[source] = { ...syncStates[source], [docId]: nextSyncState }
+
+  updatePeers(docId)
+
+  if (patch) {
+    sendPatchToFrontend({ docId, patch })
+  }
+}
+```
+
+### Applying and distributing local changes
+
+When you create a local change to a document, simply call `generateSyncMessage()` for each peer to produce a message to send them. In general, you can use the same `updatePeers()` implementation for both receiving messages or creating local changes. You may want to rate limit or debounce these communications to reduce network traffic, but this isn't required. *Remember, after applying a local change to the backend you will need to forward the resulting patch to your frontend!*
+
+Here's a sample implementation:
+
+```
+
+// Respond to messages from the frontend document
+self.addEventListener('message', (evt: any) => {
+  const { data } = evt
+  const { docId } = data
+
+  if (data.type === 'OPEN') {
+    backends[docId] = Backend.init()
+  }
+
+  if (data.type === 'LOCAL_CHANGE') {
+    const [newBackend, patch] = Backend.applyLocalChange(backends[docId], data.payload)
+    backends[docId] = newBackend
+    sendMessageToRenderer({ docId, patch })
+  }
+
+  // now tell everyone else about how things have changed
+  updatePeers(docId)
+})
+
+```
+
+### Handling disconnection
+
+Remember to save your syncState object for a peer upon disconnection via `saveSyncState()`. That might look like this:
+
+```
+db.storeSyncState(docId, source, saveSyncState(syncStates[source]))
```

Comment:
Oh I didn't realize this was required, interesting.

Reply:
It isn't required, but it will make your first few sync messages cheaper.